{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Gramene retrieval"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "address"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "from io import StringIO\n",
    "import requests\n",
    "from Bio import SeqIO"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "#gramene\n",
    "def getgramene(pid, session=None):\n",
    "    resp = (session or requests).get(\n",
    "        \"http://rest.ensembl.org/sequence/id/%s?content-type=text/x-fasta;type=protein\"% pid)\n",
    "    return SeqIO.read(StringIO(resp.text), format = \"fasta\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "#update the protein sequences in the databae\n",
    "def updatedb_prot(id_paper1, rec):\n",
    "    u = gfp_input_2019.update()\n",
    "    u = u.values({gfp_input_2019.c.seq_prot: str(rec.seq).upper()})\n",
    "    u = u.where(gfp_input_2019.c.id_paper1 == id_paper1)\n",
    "    u = u.where(gfp_input_2019.c.seq_prot == None)\n",
    "\n",
    "    proxy = engine.execute(u)\n",
    "    return proxy.rowcount\n",
    "\n",
    "def updateerr(id_paper, num):\n",
    "    u = gfp_input_2019.update()\n",
    "    u = u.values({gfp_input_2019.c.errcol: num})\n",
    "    u = u.where(gfp_input_2019.c.id_paper1 == id_paper1)\n",
    "\n",
    "    proxy = engine.execute(u)\n",
    "    return proxy.rowcount\n",
    "\n",
    "def updateresult(id_paper, source):\n",
    "    u = gfp_input_2019.update()\n",
    "    u = u.values({gfp_input_2019.c.id_type: source})\n",
    "    u = u.where(gfp_input_2019.c.id_paper1 == id_paper1)\n",
    "\n",
    "    proxy = engine.execute(u)\n",
    "    return proxy.rowcount\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### connect to database"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Set up connection to database\n",
    "from sqlalchemy import create_engine, text\n",
    "engine = create_engine('mysql+pymysql://root:PASSWORD@localhost/DATABASE')\n",
    "# test connection\n",
    "engine.execute(text(\"show variables like \\\"ver%\\\"\")).fetchall()\n",
    "\n",
    "#extract the table information\n",
    "from sqlalchemy import select, MetaData\n",
    "meta = MetaData(bind=engine)\n",
    "meta.reflect()\n",
    "meta.tables.keys()\n",
    "\n",
    "#define the table\n",
    "gfp_input_2019 = meta.tables['A_msms_input_auto1']"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Define Run variables"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "### define errors\n",
    "import time\n",
    "chunk = 100000 #how many ids in the run limit\n",
    "retrieval_delay = 2 #time break after successful retrieval in seconds\n",
    "error_delay =2 #time break after error in seconds\n",
    "max_fail=  200\n",
    "fail_num = 200\n",
    "ncbi_nuc_err = 1\n",
    "uniprot_err = 2\n",
    "ncbi_prot_err = 3\n",
    "ensembl_err = 4\n",
    "err_status = 2 #where to start sending the sequence?\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [],
   "source": [
    "###Define the nucleotide query that reoccurs after every target \n",
    "q0 = select([gfp_input_2019.c.id_paper1.distinct()])\n",
    "q1 = q0.where(gfp_input_2019.c.id_paper1!=None)\n",
    "q9 = q1.where(gfp_input_2019.c.seq_prot==None)\n",
    "q10 = q9.where(gfp_input_2019.c.seq_nucl==None)\n",
    "q20 = q10.where(gfp_input_2019.c.errcol>1 )\n",
    "qp = q20.limit(chunk)\n",
    "\n",
    "#I have to think about whether the query needs to change or how the user may want to interact with it."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "33"
      ]
     },
     "execution_count": 45,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res=[r.id_paper1 for r in engine.execute(qp).fetchall()]\n",
    "len(res)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['MusaId000049250',\n",
       " 'MusaId000042738',\n",
       " 'MusaId000032429',\n",
       " 'MusaId000030890',\n",
       " 'MusaId000027457',\n",
       " 'MusaId000027321',\n",
       " 'MusaId000019388',\n",
       " 'MusaId000019279',\n",
       " 'MusaId000018926',\n",
       " 'MusaId000018895']"
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "res[0:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Send ids of to Gramene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Sending 33  ids to Uniprot\n",
      "failed for: MusaId000049250 No records found in handle\n",
      "failed for: MusaId000042738 No records found in handle\n",
      "failed for: MusaId000032429 No records found in handle\n",
      "failed for: MusaId000030890 No records found in handle\n",
      "failed for: MusaId000027457 No records found in handle\n",
      "failed for: MusaId000027321 No records found in handle\n",
      "failed for: MusaId000019388 No records found in handle\n",
      "failed for: MusaId000019279 No records found in handle\n",
      "failed for: MusaId000018926 No records found in handle\n",
      "failed for: MusaId000018895 No records found in handle\n",
      "failed for: MusaId000018196 No records found in handle\n",
      "failed for: MusaId000017338 No records found in handle\n",
      "failed for: MusaId000016815 No records found in handle\n",
      "failed for: MusaId000011404 No records found in handle\n",
      "failed for: MusaId000010122 No records found in handle\n",
      "failed for: MusaId000010024 No records found in handle\n",
      "failed for: MusaId000009078 No records found in handle\n",
      "failed for: MusaId000008723 No records found in handle\n",
      "failed for: MusaId000007192 No records found in handle\n",
      "failed for: MusaId000006900 No records found in handle\n",
      "failed for: MusaId000006814 No records found in handle\n",
      "failed for: MusaId000006445 No records found in handle\n",
      "failed for: MusaId000006444 No records found in handle\n",
      "failed for: MusaId000004950 No records found in handle\n",
      "failed for: MusaId000004396 No records found in handle\n",
      "failed for: MusaId000004116 No records found in handle\n",
      "failed for: MusaId000003836 No records found in handle\n",
      "failed for: MusaId000003092 No records found in handle\n",
      "failed for: MusaId000003051 No records found in handle\n",
      "failed for: MusaId000002793 No records found in handle\n",
      "failed for: MusaId000002758 No records found in handle\n",
      "failed for: MusaId000002325 No records found in handle\n",
      "failed for: MusaId000000103 No records found in handle\n",
      "Done Uniprot database search!\n"
     ]
    }
   ],
   "source": [
    "print(\"Sending\",len(res),\" ids to Uniprot\")\n",
    "\n",
    "max_fail = fail_num\n",
    "for id_paper1 in res:\n",
    "    try:\n",
    "        rec = getgramene(id_paper1)\n",
    "        found = updatedb_prot(id_paper1, rec)\n",
    "        print(id_paper1, 'updated', found, 'rows')\n",
    "        time.sleep(retrieval_delay)\n",
    "        updateresult(id_paper1,\"ensemble\")\n",
    "    except Exception as e:\n",
    "        print('failed for:',id_paper1, e)\n",
    "        updateerr(id_paper1, ensembl_err)\n",
    "        max_fail -= 1\n",
    "        if max_fail <=0:\n",
    "            break\n",
    "        time.sleep(error_delay)\n",
    "print('Done Gramene database search!')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
