{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BLAST protocol\n",
    "\n",
    "We have:\n",
    "1. Primer blast: entries are primer pairs\n",
    "2. protein sequences [seq]\n",
    "3. nucleotide sequences (from NCBI)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### !!!!the script needs to run in the same folder as the databases!!!!\n",
    "\n",
    "The below final protocol has some variables and needs to loop over the different databases. The list of sequences is collected from the database for each taxaid (e.g.4081) and then send to the db with the same taxa so we have to write a function that takes in a list of taxa and runs the protocol through each. If it cannot find a taxa that was given in the list (select distinct taxaid from taxa) then skip and continue to the next id."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2 Protein sequence BLAST"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Connect to the MySQL database backing to access the sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "####DEFINE FUNCTION FOR UPDATING:\n",
    "from sqlalchemy import bindparam\n",
    "def updatedb_prot(recs):\n",
    "    u = blast_input.update()\n",
    "    u = u.values({blast_input.c.locus_seq: bindparam('subject acc.ver')})\n",
    "    u = u.where(blast_input.c.id == bindparam('query acc.ver'))\n",
    "    u = u.where(blast_input.c.locus_seq==None)\n",
    "\n",
    "    proxy = engine.execute(u, *recs)\n",
    "    return print(\"updated \",proxy.rowcount,\" entries for\",\"taxa ID\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('version', '5.7.26'),\n",
       " ('version_comment', 'MySQL Community Server (GPL)'),\n",
       " ('version_compile_machine', 'x86_64'),\n",
       " ('version_compile_os', 'macos10.14')]"
      ]
     },
     "execution_count": 38,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sqlalchemy import create_engine, text\n",
    "engine = create_engine('mysql+pymysql://root:PASSWORD@localhost/DATABASE')\n",
    "# test connection\n",
    "engine.execute(text(\"show variables like \\\"ver%\\\"\")).fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sqlalchemy import select, MetaData\n",
    "meta = MetaData(bind=engine)\n",
    "meta.reflect()\n",
    "meta.tables.keys()\n",
    "blast_input = meta.tables['test_IDseeqer']#add table you want to update\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## In one step protocol\n",
    "\n",
    "The module has to create a function for each species"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### tomato - 4081"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "updated  5  entries for taxa ID\n",
      "Updated taxa:  4081\n"
     ]
    }
   ],
   "source": [
    "#tomato 4081\n",
    "chunk = 10000\n",
    "import pandas as pd\n",
    "#####lets try to write it all in one cell:\n",
    "#select species sequences - TA\n",
    "q0 = select([blast_input.c.seq_prot,blast_input.c.id])\n",
    "q2 = q0.where(blast_input.c.locus_seq==None)\n",
    "q3 = q2.where(blast_input.c.seq_prot!=None)\n",
    "q1 = q3.where(blast_input.c.taxa==4081) #ENTER TAXA FOR PLANT HERE\n",
    "qp = q1.limit(chunk)\n",
    "rec = list(engine.execute(qp).fetchall())\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "records = [SeqRecord(id=str(r.id),seq=Seq(r.seq_prot))for r in rec] \n",
    "\n",
    "from Bio import SeqIO\n",
    "SeqIO.write(records,\"query_prot_4081.fasta\",\"fasta\")# ENTER THE species taxa here and in the blast query below (twice)!!\n",
    "#Enter the right blast database\n",
    "!blastp -max_hsps 1 -max_target_seqs 5 -outfmt 6 -query query_prot_4081.fasta -db blastdb_sl_prot -out prot_4081.tsv\n",
    "\n",
    "header='query acc.ver, subject acc.ver, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score'.split(',')\n",
    "header = [h.strip() for h in header]\n",
    "df = pd.read_csv(\"prot_4081.tsv\",header = 0, sep = \"\\t\", names=header)#ENTER THE taxa here\n",
    "\n",
    "if len(df) == 0:\n",
    "    print(\"Could not find any hits in the blast database for the \",len(rec),\" records.\")\n",
    "\n",
    "else:\n",
    "\n",
    "    dfs=df.sort_values(by = [\"query acc.ver\",\"bit score\"], ascending = [True,False])\n",
    "    idx = df.groupby(\"query acc.ver\")['bit score'].transform(max) == df['bit score']\n",
    "    dff=df[idx]\n",
    "    d=dff[['query acc.ver', 'subject acc.ver']]\n",
    "    recs=d.to_dict(orient='records')\n",
    "\n",
    "    updatedb_prot(recs)\n",
    "    print(\"Updated taxa:  4081\") #Final message with info on updated species taxa\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### potato - 4113"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### vine - 29760"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### b napus - 3708"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### soybean - 3847"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Could not find any hits in the blast database for the  4  records.\n"
     ]
    }
   ],
   "source": [
    "#3847\n",
    "chunk = 9000\n",
    "import pandas as pd\n",
    "#####lets try to write it all in one cell:\n",
    "#select species sequences - TA\n",
    "q0 = select([blast_input.c.seq_prot,blast_input.c.id])\n",
    "q2 = q0.where(blast_input.c.locus_seq==None)\n",
    "q3 = q2.where(blast_input.c.seq_prot!=None)\n",
    "q1 = q3.where(blast_input.c.taxa==3847) #ENTER TAXA FOR PLANT HERE\n",
    "qp = q1.limit(chunk)\n",
    "rec = list(engine.execute(qp).fetchall())\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "records = [SeqRecord(id=str(r.id),seq=Seq(r.seq_prot))for r in rec] \n",
    "\n",
    "from Bio import SeqIO\n",
    "SeqIO.write(records,\"query_prot_3847.fasta\",\"fasta\")# ENTER THE species taxa here and in the blast query below (twice)!!\n",
    "#Enter the right blast database\n",
    "!blastp -max_hsps 1 -max_target_seqs 5 -outfmt 6 -query query_prot_3847.fasta -db blastdb_gm_prot -out prot_3847.tsv\n",
    "header='query acc.ver, subject acc.ver, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score'.split(',')\n",
    "header = [h.strip() for h in header]\n",
    "df = pd.read_csv(\"prot_3847.tsv\",header = 0, sep = \"\\t\", names=header)#ENTER THE taxa here\n",
    "\n",
    "if len(df) == 0:\n",
    "    print(\"Could not find any hits in the blast database for the \",len(rec),\" records.\")\n",
    "\n",
    "else:\n",
    "    dfs=df.sort_values(by = [\"query acc.ver\",\"bit score\"], ascending = [True,False])\n",
    "    idx = df.groupby(\"query acc.ver\")['bit score'].transform(max) == df['bit score']\n",
    "    dff=df[idx]\n",
    "    d=dff[['query acc.ver', 'subject acc.ver']]\n",
    "    recs=d.to_dict(orient='records')\n",
    "\n",
    "    updatedb_prot(recs)\n",
    "    print(\"Updated taxa:  3847\") #Final message with info on updated species taxa\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### b rapa - 3711"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### rice - 4530"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Could not find any hits in the blast database for the  18  records.\n"
     ]
    }
   ],
   "source": [
    "# 4530\n",
    "chunk = 10000\n",
    "import pandas as pd\n",
    "#####lets try to write it all in one cell:\n",
    "#select species sequences - TA\n",
    "q0 = select([blast_input.c.seq_prot,blast_input.c.id])\n",
    "q2 = q0.where(blast_input.c.locus_seq==None)\n",
    "q3 = q2.where(blast_input.c.seq_prot!=None)\n",
    "q1 = q3.where(blast_input.c.taxa==4530) #ENTER TAXA FOR PLANT HERE\n",
    "qp = q1.limit(chunk)\n",
    "rec = list(engine.execute(qp).fetchall())\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "records = [SeqRecord(id=str(r.id),seq=Seq(r.seq_prot))for r in rec] \n",
    "\n",
    "from Bio import SeqIO\n",
    "SeqIO.write(records,\"query_prot_4530.fasta\",\"fasta\")# ENTER THE species taxa here and in the blast query below (twice)!!\n",
    "#Enter the right blast database\n",
    "!blastp -max_hsps 1 -max_target_seqs 5 -outfmt 6 -query query_prot_4530.fasta -db blastdb_os_prot -out prot_4530.tsv\n",
    "\n",
    "header='query acc.ver, subject acc.ver, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score'.split(',')\n",
    "header = [h.strip() for h in header]\n",
    "df = pd.read_csv(\"prot_4530.tsv\",header = 0, sep = \"\\t\", names=header)#ENTER THE taxa here\n",
    "\n",
    "if len(df) == 0:\n",
    "    print(\"Could not find any hits in the blast database for the \",len(rec),\" records.\")\n",
    "\n",
    "else:\n",
    "    dfs=df.sort_values(by = [\"query acc.ver\",\"bit score\"], ascending = [True,False])\n",
    "    idx = df.groupby(\"query acc.ver\")['bit score'].transform(max) == df['bit score']\n",
    "    dff=df[idx]\n",
    "    d=dff[['query acc.ver', 'subject acc.ver']]\n",
    "    recs=d.to_dict(orient='records')\n",
    "\n",
    "    updatedb_prot(recs)\n",
    "    print(\"Updated taxa:  4530\") #Final message with info on updated species taxa\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### maize 4577"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### sorghum  4558"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### banana - 4641"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### barley - 4513"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### wheat - 4565"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# STEP by STEP"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Retrieve the entries from the data table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "#####lets try to write it all in one cell:\n",
    "#select species sequences - TA\n",
    "q0 = select([blast_input.c.seq_prot,blast_input.c.id])\n",
    "q2 = q0.where(blast_input.c.locus_seq==None) #take these out to see the ones that were linked up\n",
    "q1 = q0.where(blast_input.c.taxa==\"4081\") #tomato\n",
    "qp = q1.limit(10)\n",
    "\n",
    "#generate the list of sequences and the FASTA file\n",
    "rec = list(engine.execute(qp).fetchall())"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Create the query FASTA"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "records = [SeqRecord(id=str(r.id),seq=Seq(r.seq_prot))for r in rec] \n",
    "\n",
    "from Bio import SeqIO\n",
    "SeqIO.write(records,\"query_prot_sl.fasta\",\"fasta\")\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Send it to the blast database locally\n",
    "!blastp -max_hsps 1 -max_target_seqs 5 -outfmt 6 -query query_prot_sl.fasta -db blastdb_sl_prot -out prot_ls.tsv\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make the header for the return table and create a dataframe\n",
    "header='query acc.ver, subject acc.ver, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score'.split(',')\n",
    "header = [h.strip() for h in header]\n",
    "df = pd.read_csv(\"prot_ls.tsv\",header = 0, sep = \"\\t\", names=header)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "#picking the top score\n",
    "dfs=df.sort_values(by = [\"query acc.ver\",\"bit score\"], ascending = [True,False])\n",
    "idx = df.groupby(\"query acc.ver\")['bit score'].transform(max) == df['bit score']\n",
    "dff=df[idx]\n",
    "d=dff[['query acc.ver', 'subject acc.ver']]\n",
    "recs=d.to_dict(orient='records')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[{'query acc.ver': 23664, 'subject acc.ver': 'Solyc08g062630.2.1'},\n",
       " {'query acc.ver': 23665, 'subject acc.ver': 'Solyc08g081250.2.1'},\n",
       " {'query acc.ver': 23666, 'subject acc.ver': 'Solyc08g081250.2.1'},\n",
       " {'query acc.ver': 23667, 'subject acc.ver': 'Solyc08g081250.2.1'},\n",
       " {'query acc.ver': 23668, 'subject acc.ver': 'Solyc08g081250.2.1'},\n",
       " {'query acc.ver': 23669, 'subject acc.ver': 'Solyc08g081250.2.1'},\n",
       " {'query acc.ver': 23670, 'subject acc.ver': 'Solyc08g081250.2.1'},\n",
       " {'query acc.ver': 23671, 'subject acc.ver': 'Solyc08g081250.2.1'},\n",
       " {'query acc.ver': 23672, 'subject acc.ver': 'Solyc08g081250.2.1'},\n",
       " {'query acc.ver': 23673, 'subject acc.ver': 'Solyc08g081250.2.1'}]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "recs"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Bind the top scoring hit to the id in the retrieval table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sqlalchemy import bindparam\n",
    "def updatedb_prot(recs):\n",
    "    u = blast_input.update()\n",
    "    u = u.values({blast_input.c.locus_seq: bindparam('subject acc.ver')})\n",
    "    u = u.where(blast_input.c.id == bindparam('query acc.ver'))\n",
    "    #u = u.where(blast_input.c.locus_seq==None)\n",
    "\n",
    "    proxy = engine.execute(u, *recs)\n",
    "    return proxy.rowcount\n",
    "\n",
    "updatedb_prot(recs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
