{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# FASTA download from Ensembl Server"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Download the FASTA files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The server provides the same structure to find the files. Maybe we can automate the retrieval where:\n",
    "1. the first script scrapes the names<br>\n",
    "    Find server address\n",
    "    ftp://ftp.ensemblgenomes.org/pub/release-40/plants/fasta/arabidopsis_thaliana/cdna/<br>\n",
    "    The variables in this are:<br>\n",
    "    RELEASE: release-40 (new releases will have higher numbers e.g. current release = 44)<br>\n",
    "    DATA: cdna or pep <br>\n",
    "    SPECIES: appear at two points in the address and can be optained from the `taxa` table `species_name` <br>\n",
    "    Then scrape the name of the file from the server page that ends with *.cdna.all.fa.gz or *.pep.all.fa.gz respecively\n",
    "\n",
    "\n",
    "2. the second script generatea the curl query (should look like below)\n",
    "\n",
    "\n",
    "\n",
    "3. the third script executes them\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 20.2M  100 20.2M    0     0  1316k      0  0:00:15  0:00:15 --:--:-- 3828k 0   8848      0  0:40:04  0:00:05  0:39:59 10751\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 39.9M  100 39.9M    0     0  2307k      0  0:00:17  0:00:17 --:--:-- 6713k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 39.9M  100 39.9M    0     0  1817k      0  0:00:22  0:00:22 --:--:-- 4485k  53849      0  0:12:57  0:00:08  0:12:49 910960   191k      0  0:03:33  0:00:11  0:03:22  424k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 17.7M  100 17.7M    0     0  1125k      0  0:00:16  0:00:16 --:--:-- 3472k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 13.2M  100 13.2M    0     0   740k      0  0:00:18  0:00:18 --:--:-- 2410k0     0      0      0 --:--:--  0:00:01 --:--:--     0 0  0:01:26  0:00:12  0:01:14  335k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 20.2M  100 20.2M    0     0   719k      0  0:00:28  0:00:28 --:--:-- 2779k33      0  0:23:01  0:00:06  0:22:55 19491\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 13.0M  100 13.0M    0     0   600k      0  0:00:22  0:00:22 --:--:-- 1556k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 21.1M  100 21.1M    0     0   979k      0  0:00:22  0:00:22 --:--:-- 2983k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 52.5M  100 52.5M    0     0  1691k      0  0:00:31  0:00:31 --:--:-- 4484k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 25.6M  100 25.6M    0     0  1323k      0  0:00:19  0:00:19 --:--:-- 4142k  772k      0  0:00:33  0:00:17  0:00:16 2207k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 15.0M  100 15.0M    0     0   621k      0  0:00:24  0:00:24 --:--:-- 1551k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 66.8M  100 66.8M    0     0  1761k      0  0:00:38  0:00:38 --:--:-- 6175k\n",
      "  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current\n",
      "                                 Dload  Upload   Total   Spent    Left  Speed\n",
      "100 49.9M  100 49.9M    0     0  1934k      0  0:00:26  0:00:26 --:--:-- 4872k\n"
     ]
    }
   ],
   "source": [
    "#download the file for the protein search\n",
    "!curl -o Arabidopsis_thaliana.TAIR10.cdna.all.fa.gz ftp://ftp.ensemblgenomes.org/pub/release-40/plants/fasta/arabidopsis_thaliana/cdna/Arabidopsis_thaliana.TAIR10.cdna.all.fa.gz\n",
    "!curl -o Glycine_max.Glycine_max_v2.0.cdna.all.fa.gz ftp://ftp.ensemblgenomes.org/pub/release-40/plants/fasta/glycine_max/cdna/Glycine_max.Glycine_max_v2.0.cdna.all.fa.gz\n",
    "!curl -o Brassica_napus.AST_PRJEB5043_v1.cdna.all.fa.gz ftp://ftp.ensemblgenomes.org/pub/release-40/plants/fasta/brassica_napus/cdna/Brassica_napus.AST_PRJEB5043_v1.cdna.all.fa.gz\n",
    "!curl -o Brassica_rapa.IVFCAASv1.cdna.all.fa.gz ftp://ftp.ensemblgenomes.org/pub/release-40/plants/fasta/brassica_rapa/cdna/Brassica_rapa.IVFCAASv1.cdna.all.fa.gz\n",
    "!curl -o Solanum_lycopersicum.SL2.50.cdna.all.fa.gz ftp://ftp.ensemblgenomes.org/pub/release-40/plants/fasta/solanum_lycopersicum/cdna/Solanum_lycopersicum.SL2.50.cdna.all.fa.gz\n",
    "!curl -o Solanum_tuberosum.SolTub_3.0.cdna.all.fa.gz ftp://ftp.ensemblgenomes.org/pub/release-40/plants/fasta/solanum_tuberosum/cdna/Solanum_tuberosum.SolTub_3.0.cdna.all.fa.gz\n",
    "!curl -o Vitis_vinifera.IGGP_12x.cdna.all.fa.gz ftp://ftp.ensemblgenomes.org/pub/release-40/plants/fasta/vitis_vinifera/cdna/Vitis_vinifera.IGGP_12x.cdna.all.fa.gz\n",
    "!curl -o Oryza_sativa.IRGSP-1.0.cdna.all.fa.gz ftp://ftp.ensemblgenomes.org/pub/release-40/plants/fasta/oryza_sativa/cdna/Oryza_sativa.IRGSP-1.0.cdna.all.fa.gz\n",
    "!curl -o Zea_mays.AGPv4.cdna.all.fa.gz ftp://ftp.ensemblgenomes.org/pub/release-40/plants/fasta/zea_mays/cdna/Zea_mays.AGPv4.cdna.all.fa.gz\n",
    "!curl -o Sorghum_bicolor.Sorghum_bicolor_NCBIv3.cdna.all.fa.gz ftp://ftp.ensemblgenomes.org/pub/release-40/plants/fasta/sorghum_bicolor/cdna/Sorghum_bicolor.Sorghum_bicolor_NCBIv3.cdna.all.fa.gz\n",
    "!curl -o Musa_acuminata.MA1.cdna.all.fa.gz ftp://ftp.ensemblgenomes.org/pub/release-40/plants/fasta/musa_acuminata/cdna/Musa_acuminata.MA1.cdna.all.fa.gz\n",
    "!curl -o Hordeum_vulgare.Hv_IBSC_PGSB_v2.cdna.all.fa.gz ftp://ftp.ensemblgenomes.org/pub/release-40/plants/fasta/hordeum_vulgare/cdna/Hordeum_vulgare.Hv_IBSC_PGSB_v2.cdna.all.fa.gz\n",
    "!curl -o Triticum_aestivum.IWGSC.cdna.all.fa.gz ftp://ftp.ensemblgenomes.org/pub/release-40/plants/fasta/triticum_aestivum/cdna/Triticum_aestivum.IWGSC.cdna.all.fa.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Build the Databases\n",
    "\n",
    "The cDNA files are then collected and converted to a species database that can be BLASTED (see blast script)\n",
    "\n",
    "This can be done automatically similar to the above, where the `taxa` table is used to generate the query using the cdna files and filling in the taxa numbers into the BLASTdb names: blastdb_3704_nucl as well as the type of data:\n",
    "\n",
    "cdna: nucl in the name so we know which database it is and -dbtype = nucl so the computer knows which one it is.<br>\n",
    "pep: prot in the name so we know which database it is and -dbtype = prot so the computer knows which one it is.\n",
    "\n",
    "!gzcat Arabidopsis_thaliana.TAIR10.cdna.all.fa.gz |makeblastdb -in - -out blastdb_3704_nucl -input_type=fasta -dbtype nucl -title athaliana\n",
    "\n",
    "-title: can extract the `species_name` from `taxa`. This is only informative and not rigit\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "Building a new DB, current time: 08/05/2019 10:09:50\n",
      "New DB name:   /Users/connyhooper/Dropbox/12_Database_projects/6_cropPAL-2/cropPAL2v2_built_20181109/BLAST_ID_retrieval/FASTA_download/cdna/blastdb_at_nucl\n",
      "New DB title:  athaliana\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 48359 sequences in 7.24462 seconds.\n",
      "\n",
      "\n",
      "Building a new DB, current time: 08/05/2019 10:09:58\n",
      "New DB name:   /Users/connyhooper/Dropbox/12_Database_projects/6_cropPAL-2/cropPAL2v2_built_20181109/BLAST_ID_retrieval/FASTA_download/cdna/blastdb_gm_nucl\n",
      "New DB title:  gmax\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 89597 sequences in 14.3019 seconds.\n",
      "\n",
      "\n",
      "Building a new DB, current time: 08/05/2019 10:10:12\n",
      "New DB name:   /Users/connyhooper/Dropbox/12_Database_projects/6_cropPAL-2/cropPAL2v2_built_20181109/BLAST_ID_retrieval/FASTA_download/cdna/blastdb_bn_nucl\n",
      "New DB title:  bnapus\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 101040 sequences in 11.84 seconds.\n",
      "\n",
      "\n",
      "Building a new DB, current time: 08/05/2019 10:10:24\n",
      "New DB name:   /Users/connyhooper/Dropbox/12_Database_projects/6_cropPAL-2/cropPAL2v2_built_20181109/BLAST_ID_retrieval/FASTA_download/cdna/blastdb_br_nucl\n",
      "New DB title:  Brava\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 41025 sequences in 4.93979 seconds.\n",
      "\n",
      "\n",
      "Building a new DB, current time: 08/05/2019 10:10:29\n",
      "New DB name:   /Users/connyhooper/Dropbox/12_Database_projects/6_cropPAL-2/cropPAL2v2_built_20181109/BLAST_ID_retrieval/FASTA_download/cdna/blastdb_sl_nucl\n",
      "New DB title:  slycopersicum\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 33810 sequences in 3.89271 seconds.\n",
      "\n",
      "\n",
      "Building a new DB, current time: 08/05/2019 10:10:33\n",
      "New DB name:   /Users/connyhooper/Dropbox/12_Database_projects/6_cropPAL-2/cropPAL2v2_built_20181109/BLAST_ID_retrieval/FASTA_download/cdna/blastdb_st_nucl\n",
      "New DB title:  stuberosum\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 56210 sequences in 7.23003 seconds.\n",
      "\n",
      "\n",
      "Building a new DB, current time: 08/05/2019 10:10:41\n",
      "New DB name:   /Users/connyhooper/Dropbox/12_Database_projects/6_cropPAL-2/cropPAL2v2_built_20181109/BLAST_ID_retrieval/FASTA_download/cdna/blastdb_vv_nucl\n",
      "New DB title:  vvinifera\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 29971 sequences in 3.68091 seconds.\n",
      "\n",
      "\n",
      "Building a new DB, current time: 08/05/2019 10:10:44\n",
      "New DB name:   /Users/connyhooper/Dropbox/12_Database_projects/6_cropPAL-2/cropPAL2v2_built_20181109/BLAST_ID_retrieval/FASTA_download/cdna/blastdb_os_nucl\n",
      "New DB title:  osativa\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 42387 sequences in 5.91095 seconds.\n",
      "\n",
      "\n",
      "Building a new DB, current time: 08/05/2019 10:10:50\n",
      "New DB name:   /Users/connyhooper/Dropbox/12_Database_projects/6_cropPAL-2/cropPAL2v2_built_20181109/BLAST_ID_retrieval/FASTA_download/cdna/blastdb_zm_nucl\n",
      "New DB title:  mays\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 131496 sequences in 25.1279 seconds.\n",
      "\n",
      "\n",
      "Building a new DB, current time: 08/05/2019 10:11:16\n",
      "New DB name:   /Users/connyhooper/Dropbox/12_Database_projects/6_cropPAL-2/cropPAL2v2_built_20181109/BLAST_ID_retrieval/FASTA_download/cdna/blastdb_sb_nucl\n",
      "New DB title:  bicolor\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 47110 sequences in 8.29155 seconds.\n",
      "\n",
      "\n",
      "Building a new DB, current time: 08/05/2019 10:11:24\n",
      "New DB name:   /Users/connyhooper/Dropbox/12_Database_projects/6_cropPAL-2/cropPAL2v2_built_20181109/BLAST_ID_retrieval/FASTA_download/cdna/blastdb_ma_nucl\n",
      "New DB title:  macuminata\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 36525 sequences in 4.61063 seconds.\n",
      "\n",
      "\n",
      "Building a new DB, current time: 08/05/2019 10:11:29\n",
      "New DB name:   /Users/connyhooper/Dropbox/12_Database_projects/6_cropPAL-2/cropPAL2v2_built_20181109/BLAST_ID_retrieval/FASTA_download/cdna/blastdb_hv_nucl\n",
      "New DB title:  hvulgare\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 248391 sequences in 37.2938 seconds.\n",
      "\n",
      "\n",
      "Building a new DB, current time: 08/05/2019 10:12:06\n",
      "New DB name:   /Users/connyhooper/Dropbox/12_Database_projects/6_cropPAL-2/cropPAL2v2_built_20181109/BLAST_ID_retrieval/FASTA_download/cdna/blastdb_ta_nucl\n",
      "New DB title:  taestivum\n",
      "Sequence type: Nucleotide\n",
      "Keep MBits: T\n",
      "Maximum file size: 1000000000B\n",
      "Adding sequences from FASTA; added 137056 sequences in 18.3389 seconds.\n"
     ]
    }
   ],
   "source": [
    "#BLAST nuclein DATABASE BUILD\n",
    "!gzcat Arabidopsis_thaliana.TAIR10.cdna.all.fa.gz |makeblastdb -in - -out blastdb_at_nucl -input_type=fasta -dbtype nucl -title athaliana\n",
    "!gzcat Glycine_max.Glycine_max_v2.0.cdna.all.fa.gz |makeblastdb -in - -out blastdb_gm_nucl -input_type=fasta -dbtype nucl -title gmax\n",
    "!gzcat Brassica_napus.AST_PRJEB5043_v1.cdna.all.fa.gz |makeblastdb -in - -out blastdb_bn_nucl -input_type=fasta -dbtype nucl -title bnapus\n",
    "!gzcat Brassica_rapa.IVFCAASv1.cdna.all.fa.gz |makeblastdb -in - -out blastdb_br_nucl -input_type=fasta -dbtype nucl -title brapa\n",
    "!gzcat Solanum_lycopersicum.SL2.50.cdna.all.fa.gz |makeblastdb -in - -out blastdb_sl_nucl -input_type=fasta -dbtype nucl -title slycopersicum\n",
    "!gzcat Solanum_tuberosum.SolTub_3.0.cdna.all.fa.gz|makeblastdb -in - -out blastdb_st_nucl -input_type=fasta -dbtype nucl -title stuberosum\n",
    "!gzcat Vitis_vinifera.IGGP_12x.cdna.all.fa.gz |makeblastdb -in - -out blastdb_vv_nucl -input_type=fasta -dbtype nucl -title vvinifera\n",
    "\n",
    "!gzcat Oryza_sativa.IRGSP-1.0.cdna.all.fa.gz |makeblastdb -in - -out blastdb_os_nucl -input_type=fasta -dbtype nucl -title osativa\n",
    "!gzcat Zea_mays.AGPv4.cdna.all.fa.gz |makeblastdb -in - -out blastdb_zm_nucl -input_type=fasta -dbtype nucl -title mays\n",
    "!gzcat Sorghum_bicolor.Sorghum_bicolor_NCBIv3.cdna.all.fa.gz |makeblastdb -in - -out blastdb_sb_nucl -input_type=fasta -dbtype nucl -title bicolor\n",
    "!gzcat Musa_acuminata.MA1.cdna.all.fa.gz |makeblastdb -in - -out blastdb_ma_nucl -input_type=fasta -dbtype nucl -title macuminata\n",
    "!gzcat Hordeum_vulgare.Hv_IBSC_PGSB_v2.cdna.all.fa.gz |makeblastdb -in - -out blastdb_hv_nucl -input_type=fasta -dbtype nucl -title hvulgare\n",
    "!gzcat Triticum_aestivum.IWGSC.cdna.all.fa.gz |makeblastdb -in - -out blastdb_ta_nucl -input_type=fasta -dbtype nucl -title taestivum\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### NOTES\n",
    "1. The labelling of the files needs to be by `taxaid` (e.g. 4565 etc etc). The above doesn't do that. It still uses initials of the species name. That needs to be changed using the `taxa` table in the database. This may requires a sql script taking the `taxa` table column `species_name`.\n",
    "2. The whole thing has to run twice: once to download `cdna` and then make the `nucl` databases, then download `pep` and make the `prot` databases"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
