{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# BLAST protocol\n",
    "\n",
    "We have:\n",
    "1. Primer blast: entries are primer pairs\n",
    "2. protein sequences [seq]  (see Blast_cropPAL_prot)\n",
    "3. nucleotide sequences (from NCBI)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### !!!!the script needs to run in the same folder as the databases!!!!\n",
    "\n",
    "#### the nucleotide script will run last and only runs the ones that have no locus_seq entry yet (haven't been matched by the protein BLAST yet)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 2 Nucleotide sequence BLAST"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Connect to the MySQL database backing to access the sequences"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "####DEFINE FUNCTION FOR UPDATING:\n",
    "from sqlalchemy import bindparam\n",
    "def updatedb_nucl(recs):\n",
    "    u = blast_input.update()\n",
    "    u = u.values({blast_input.c.locus_seq: bindparam('subject acc.ver')})\n",
    "    u = u.where(blast_input.c.id == bindparam('query acc.ver'))\n",
    "    u = u.where(blast_input.c.locus_seq==None)\n",
    "\n",
    "    proxy = engine.execute(u, *recs)\n",
    "    return print(\"updated \",proxy.rowcount,\" entries for\",\"taxa ID here\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('version', '5.7.26'),\n",
       " ('version_comment', 'MySQL Community Server (GPL)'),\n",
       " ('version_compile_machine', 'x86_64'),\n",
       " ('version_compile_os', 'macos10.14')]"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "from sqlalchemy import create_engine, text\n",
    "engine = create_engine('mysql+pymysql://root:PASSWORD@localhost/DATABASE')\n",
    "# test connection\n",
    "engine.execute(text(\"show variables like \\\"ver%\\\"\")).fetchall()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [],
   "source": [
    "from sqlalchemy import select, MetaData\n",
    "meta = MetaData(bind=engine)\n",
    "meta.reflect()\n",
    "meta.tables.keys()\n",
    "blast_input = meta.tables['test_IDseeqer_temp']#put datatable template there\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## In one step protocol"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### tomato - 4081"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "updated  17  entries for taxa ID here\n",
      "Updated taxa:  4081\n"
     ]
    }
   ],
   "source": [
    "#tomato 4081\n",
    "chunk = 1000\n",
    "import pandas as pd\n",
    "#####lets try to write it all in one cell:\n",
    "#select species sequences - TA\n",
    "q0 = select([blast_input.c.seq_nucl,blast_input.c.id])\n",
    "q2 = q0.where(blast_input.c.locus_seq==None)\n",
    "q3 = q2.where(blast_input.c.seq_nucl!=None)\n",
    "q1 = q3.where(blast_input.c.taxa==4081) #ENTER TAXA FOR PLANT HERE\n",
    "qn = q1.limit(chunk)\n",
    "rec = list(engine.execute(qn).fetchall())\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "records = [SeqRecord(id=str(r.id),seq=Seq(r.seq_nucl))for r in rec] \n",
    "\n",
    "from Bio import SeqIO\n",
    "SeqIO.write(records,\"query_nucl_4081.fasta\",\"fasta\")# ENTER THE species taxa here and in the blast query below (twice)!!\n",
    "#Enter the right blast database\n",
    "!blastn -max_hsps 1 -max_target_seqs 5 -outfmt 6 -query query_nucl_4081.fasta -db blastdb_sl_nucl -out nucl_4081.tsv -task blastn-short\n",
    "\n",
    "header='query acc.ver, subject acc.ver, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score'.split(',')\n",
    "header = [h.strip() for h in header]\n",
    "df = pd.read_csv(\"nucl_4081.tsv\",header = 0, sep = \"\\t\", names=header)#ENTER THE taxa here\n",
    "\n",
    "if len(df) == 0:\n",
    "    print(\"Could not find any hits in the blast database for the \",len(rec),\" records.\")\n",
    "\n",
    "else:\n",
    "    dfs=df.sort_values(by = [\"query acc.ver\",\"bit score\"], ascending = [True,False])\n",
    "    idx = df.groupby(\"query acc.ver\")['bit score'].transform(max) == df['bit score']\n",
    "    dff=df[idx]\n",
    "    d=dff[['query acc.ver', 'subject acc.ver']]\n",
    "    recs=d.to_dict(orient='records')\n",
    "\n",
    "    updatedb_nucl(recs)\n",
    "    print(\"Updated taxa:  4081\") #Final message with info on updated species taxa\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### potato - 4113"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "updated  1  entries for taxa ID here\n",
      "Updated taxa:  4113\n"
     ]
    }
   ],
   "source": [
    "#tomato 4113\n",
    "chunk = 1000\n",
    "import pandas as pd\n",
    "#####lets try to write it all in one cell:\n",
    "#select species sequences - TA\n",
    "q0 = select([blast_input.c.seq_nucl,blast_input.c.id])\n",
    "q2 = q0.where(blast_input.c.locus_seq==None)\n",
    "q3 = q2.where(blast_input.c.seq_nucl!=None)\n",
    "q1 = q3.where(blast_input.c.taxa==4113) #ENTER TAXA FOR PLANT HERE\n",
    "qn = q1.limit(chunk)\n",
    "rec = list(engine.execute(qn).fetchall())\n",
    "print(len(rec))\n",
    "\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "records = [SeqRecord(id=str(r.id),seq=Seq(r.seq_nucl))for r in rec] \n",
    "\n",
    "from Bio import SeqIO\n",
    "SeqIO.write(records,\"query_nucl_4113.fasta\",\"fasta\")# ENTER THE species taxa here and in the blast query below (twice)!!\n",
    "#Enter the right blast database\n",
    "!blastn -max_hsps 1 -max_target_seqs 5 -outfmt 6 -query query_nucl_4113.fasta -db blastdb_st_nucl -out nucl_4113.tsv  -task blastn-short\n",
    "\n",
    "header='query acc.ver, subject acc.ver, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score'.split(',')\n",
    "header = [h.strip() for h in header]\n",
    "df = pd.read_csv(\"nucl_4113.tsv\",header = 0, sep = \"\\t\", names=header)#ENTER THE taxa here\n",
    "\n",
    "\n",
    "if len(df) == 0:\n",
    "    print(\"Could not find any hits in the blast database for the \",len(rec),\" records.\")\n",
    "\n",
    "else:\n",
    "    dfs=df.sort_values(by = [\"query acc.ver\",\"bit score\"], ascending = [True,False])\n",
    "    idx = df.groupby(\"query acc.ver\")['bit score'].transform(max) == df['bit score']\n",
    "    dff=df[idx]\n",
    "    d=dff[['query acc.ver', 'subject acc.ver']]\n",
    "    recs=d.to_dict(orient='records')\n",
    "\n",
    "    updatedb_nucl(recs)\n",
    "    print(\"Updated taxa:  4113\") #Final message with info on updated species taxa\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### vine - 29760"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "Warning: [blastn] Query is Empty!\n",
      "Could not find any hits in the blast database for the  0  records.\n"
     ]
    }
   ],
   "source": [
    "#tomato 29760\n",
    "chunk = 1000\n",
    "import pandas as pd\n",
    "#####lets try to write it all in one cell:\n",
    "#select species sequences - TA\n",
    "q0 = select([blast_input.c.seq_nucl,blast_input.c.id])\n",
    "q2 = q0.where(blast_input.c.locus_seq==None)\n",
    "q3 = q2.where(blast_input.c.seq_nucl!=None)\n",
    "q1 = q3.where(blast_input.c.taxa==29760) #ENTER TAXA FOR PLANT HERE\n",
    "qn = q1.limit(chunk)\n",
    "rec = list(engine.execute(qn).fetchall())\n",
    "print(len(rec))\n",
    "\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "records = [SeqRecord(id=str(r.id),seq=Seq(r.seq_nucl))for r in rec] \n",
    "\n",
    "from Bio import SeqIO\n",
    "SeqIO.write(records,\"query_nucl_29760.fasta\",\"fasta\")# ENTER THE species taxa here and in the blast query below (twice)!!\n",
    "#Enter the right blast database\n",
    "!blastn -max_hsps 1 -max_target_seqs 5 -outfmt 6 -query query_nucl_29760.fasta -db blastdb_vv_nucl -out nucl_29760.tsv  #-task blastn-short\n",
    "\n",
    "header='query acc.ver, subject acc.ver, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score'.split(',')\n",
    "header = [h.strip() for h in header]\n",
    "df = pd.read_csv(\"nucl_29760.tsv\",header = 0, sep = \"\\t\", names=header)#ENTER THE taxa here\n",
    "\n",
    "\n",
    "if len(df) == 0:\n",
    "    print(\"Could not find any hits in the blast database for the \",len(rec),\" records.\")\n",
    "\n",
    "else:\n",
    "    dfs=df.sort_values(by = [\"query acc.ver\",\"bit score\"], ascending = [True,False])\n",
    "    idx = df.groupby(\"query acc.ver\")['bit score'].transform(max) == df['bit score']\n",
    "    dff=df[idx]\n",
    "    d=dff[['query acc.ver', 'subject acc.ver']]\n",
    "    recs=d.to_dict(orient='records')\n",
    "\n",
    "    updatedb_nucl(recs)\n",
    "    print(\"Updated taxa:  29760\") #Final message with info on updated species taxa\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### b napus - 3708"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "updated  1  entries for taxa ID here\n",
      "Updated taxa:  3708\n"
     ]
    }
   ],
   "source": [
    "# 3708\n",
    "chunk = 1000\n",
    "import pandas as pd\n",
    "#####lets try to write it all in one cell:\n",
    "#select species sequences - TA\n",
    "q0 = select([blast_input.c.seq_nucl,blast_input.c.id])\n",
    "q2 = q0.where(blast_input.c.locus_seq==None)\n",
    "q3 = q2.where(blast_input.c.seq_nucl!=None)\n",
    "q1 = q3.where(blast_input.c.taxa==3708) #ENTER TAXA FOR PLANT HERE\n",
    "qn = q1.limit(chunk)\n",
    "rec = list(engine.execute(qn).fetchall())\n",
    "print(len(rec))\n",
    "\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "records = [SeqRecord(id=str(r.id),seq=Seq(r.seq_nucl))for r in rec] \n",
    "\n",
    "from Bio import SeqIO\n",
    "SeqIO.write(records,\"query_nucl_3708.fasta\",\"fasta\")# ENTER THE species taxa here and in the blast query below (twice)!!\n",
    "#Enter the right blast database\n",
    "!blastn -max_hsps 1 -max_target_seqs 5 -outfmt 6 -query query_nucl_3708.fasta -db blastdb_bn_nucl -out nucl_3708.tsv  -task blastn-short\n",
    "\n",
    "header='query acc.ver, subject acc.ver, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score'.split(',')\n",
    "header = [h.strip() for h in header]\n",
    "df = pd.read_csv(\"nucl_3708.tsv\",header = 0, sep = \"\\t\", names=header)#ENTER THE taxa here\n",
    "\n",
    "\n",
    "if len(df) == 0:\n",
    "    print(\"Could not find any hits in the blast database for the \",len(rec),\" records.\")\n",
    "\n",
    "else:\n",
    "    dfs=df.sort_values(by = [\"query acc.ver\",\"bit score\"], ascending = [True,False])\n",
    "    idx = df.groupby(\"query acc.ver\")['bit score'].transform(max) == df['bit score']\n",
    "    dff=df[idx]\n",
    "    d=dff[['query acc.ver', 'subject acc.ver']]\n",
    "    recs=d.to_dict(orient='records')\n",
    "\n",
    "    updatedb_nucl(recs)\n",
    "    print(\"Updated taxa:  3708\") #Final message with info on updated species taxa\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "###### soybean - 3847"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "Warning: [blastn] Query is Empty!\n",
      "Could not find any hits in the blast database for the  0  records.\n"
     ]
    }
   ],
   "source": [
    "#3847\n",
    "chunk = 1000\n",
    "import pandas as pd\n",
    "#####lets try to write it all in one cell:\n",
    "#select species sequences - TA\n",
    "q0 = select([blast_input.c.seq_nucl,blast_input.c.id])\n",
    "q2 = q0.where(blast_input.c.locus_seq==None)\n",
    "q3 = q2.where(blast_input.c.seq_nucl!=None)\n",
    "q1 = q3.where(blast_input.c.taxa==3847) #ENTER TAXA FOR PLANT HERE\n",
    "qn = q1.limit(chunk)\n",
    "rec = list(engine.execute(qn).fetchall())\n",
    "print(len(rec))\n",
    "\n",
    "\n",
    "from Bio.Seq import Seq\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "records = [SeqRecord(id=str(r.id),seq=Seq(r.seq_nucl))for r in rec] \n",
    "\n",
    "from Bio import SeqIO\n",
    "SeqIO.write(records,\"query_nucl_3847.fasta\",\"fasta\")# ENTER THE species taxa here and in the blast query below (twice)!!\n",
    "#Enter the right blast database\n",
    "!blastn -max_hsps 1 -max_target_seqs 5 -outfmt 6 -query query_nucl_3847.fasta -db blastdb_gm_nucl -out nucl_3847.tsv  -task blastn-short\n",
    "\n",
    "header='query acc.ver, subject acc.ver, % identity, alignment length, mismatches, gap opens, q. start, q. end, s. start, s. end, evalue, bit score'.split(',')\n",
    "header = [h.strip() for h in header]\n",
    "df = pd.read_csv(\"nucl_3847.tsv\",header = 0, sep = \"\\t\", names=header)#ENTER THE taxa here\n",
    "\n",
    "\n",
    "if len(df) == 0:\n",
    "    print(\"Could not find any hits in the blast database for the \",len(rec),\" records.\")\n",
    "\n",
    "else:\n",
    "    dfs=df.sort_values(by = [\"query acc.ver\",\"bit score\"], ascending = [True,False])\n",
    "    idx = df.groupby(\"query acc.ver\")['bit score'].transform(max) == df['bit score']\n",
    "    dff=df[idx]\n",
    "    d=dff[['query acc.ver', 'subject acc.ver']]\n",
    "    recs=d.to_dict(orient='records')\n",
    "\n",
    "    updatedb_nucl(recs)\n",
    "    print(\"Updated taxa:  3847\") #Final message with info on updated species taxa\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
